python_interface/design/03_pyHorace_description.md

```diff
@@ -0,0 +1,269 @@
+and not by the user calling those functions/classes from Python.
+Given this major limitation, the approach taken by the "Hugo" prototype is to use a gateway function,
+`call_method` which takes the *name* of a Matlab function, class or class method and a set of arguments and uses
+`feval` to call those "included" functions/class/methods.
+This is described in more detail in the [next section](#gateway-functions-and-proxy-classes).
+
+Point 2 means that the native Matlab way of calling Python (using the `py.*` namespace) cannot be used
+but also means that a `mex` file can be used for this purpose.
+This is described in detail in the [section on Python user functions](#calling-python-from-compiled-matlab).
+
+Point 3 means that if we would like users to be able to use `help(sqw)` or similar
+we would need to generate the help text and wrappers for each function / class we export (not currently implemented).
+This and point 4 and other limitations are discussed in the [limitations section](#limitations-and-future-work)
+
+
+# Gateway functions and proxy classes
+
+["pyHorace"](https://github.com/mducle/hugo) is based on a framework (["pySpinW"](https://github.com/spinw/pyspinw)) written for SpinW by S. Ward.
+The core of this is a Matlab "gateway" function, [`call_method.m`](https://github.com/mducle/hugo/blob/master/src/call_method.m#L101)
+which takes the name of a Matlab function and its arguments and uses `feval` to evaluate it.
+When compiled using the Matlab Compiler SDK as a Python module, it maybe accessed in Python using `<module_name>.call_method('function_name', object, [<arg_list>])`.
+To make it easier to use, this core gateway function and the Matlab Python module (called `horace`) is wrapped in a Python class,
+[`pyHorace.Matlab`](https://github.com/mducle/hugo/blob/master/pyHorace/Matlab.py#L60) whose `__getattr__()` special method is overloaded to call `call_method`.
+That is, when the user does:
+
+```
+from pyHorace import Matlab
+Matlab.<func_name>(*args)
+```
+
+The `Matlab.<func_name>` part causes Python to call `Matlab.__getattr__('func_name')`. 
+This returns a Python function which wraps `horace.call_method('func_name', [], *args)`.
+Both these steps are chained by Python so that uses can do `Matlab.cut_sqw('sqw_file', [], [], [], [-1,1])` transparently.
+Because the wrapper to `call_method` is through the `__getattr__` overload of a class, users always have to use the "dot" method-access notation.
+I.e., `from pyHorace.Matlab import *; cut_sqw('sqw_file', [], [], [], [-1,1])` will not work because `pyHorace.Matlab` is not a Python module but a class.
+In addition `help(pyHorace.Matlab.cut_sqw)` will also not work because this actually refers to a (dynamically constructed) Python wrapper function.
+Thus if we want the above two functionalities (accessing Matlab "included" functions/classes and help)
+we will need to write explicit Python wrappers for *every* Matlab function we want to expose this way.
```

Comment:
Does that extend to wrappers for MATLAB class methods?

Reply:
Yes, I think so - at least if we want to fully exploit the `pydoc` system for online help. 

If not, then it depends on whether we want to extend the syntax of that method. If the Matlab syntax accepts "keyword" arguments - (string,value) pairs - then we won't need to wrap the methods. 

I'm hoping we can just define a kind of specification and then autogenerate the wrappers...